src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -103,21 +104,27 @@ public ImmutableArray<PackageSource> GetPackageSources()
                 return packageSources;
             }
 
-            try
+            lock (_gate)
             {
-                packageSources = _packageSourceProvider.Value.GetSources(includeUnOfficial: true, includeDisabled: false)
-                    .SelectAsArray(r => new PackageSource(r.Key, r.Value));
+                if (_packageSourcesAsync is null)
+                {
+                    _packageSourcesAsync = ThreadingContext.JoinableTaskFactory.RunAsync(() => GetPackageSourcesImplAsync());
+                }
             }
-            catch (Exception ex) when (ex is InvalidDataException || ex is InvalidOperationException)
+
+            if (allowSwitchToMainThread)
             {
-                // These exceptions can happen when the nuget.config file is broken.
-                packageSources = ImmutableArray<PackageSource>.Empty;
+                using var combinedToken = _tokenSource.Token.CombineWith(cancellationToken);
+                packageSources = await _packageSourcesAsync.JoinAsync(combinedToken.Token).ConfigureAwait(false);
             }
-            catch (ArgumentException ae) when (FatalError.ReportWithoutCrash(ae))
+            else if (_packageSourcesAsync.IsCompleted)
             {
-                // This exception can happen when the nuget.config file is broken, e.g. invalid credentials.
-                // https://github.com/dotnet/roslyn/issues/40857
-                packageSources = ImmutableArray<PackageSource>.Empty;
+                packageSources = _packageSourcesAsync.GetAwaiter().GetResult();
+            }
+            else
+            {
+                // Return without caching a result
```

Comment:
doc the impact.  primarily that some features may show incomplete resulsts for a short period of time, but that that's acceptable.

Reply:
Done